File: go/general/137. 只出现一次的数字 II.go
package main

func singleNumber(nums []int) int {
	res := 0
	for i := 0; i < 64; i++ {
		cnt := 0      // 记录当前bit位有多少个1
		bit := 1 << i // 记录当前要操作的bit
		for _, num := range nums {
			if (num & bit) != 0 {
				cnt += 1
			}
		}
		if cnt%3 != 0 { // 不等于0说明唯一的数字在这个bit上是1
			res |= bit
		}
	}
	return res
}

File: python/general/137. 只出现一次的数字 II.py
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        res = 0
        for i in range(32):
            cnt = 0  # 记录当前 bit 有多少个1
            bit = 1 << i  # 记录当前要操作的 bit
            for num in nums:
                if num & bit != 0:
                    cnt += 1
            if cnt % 3 != 0:
                # 不等于0说明唯一出现的数字在这个 bit 上是1
                res |= bit

        return res - 2 ** 32 if res > 2 ** 31 - 1 else res


# 数学法
# [A,A,A,B,B,B,C,C,C]和[A,A,A,B,B,B,C]差了两个C
# 3*(A+B+C) - 3*(A+B) = 2*C
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        return (sum(set(nums))*3 - sum(nums))//2

